crates/transport/src/error.rs

```diff
@@ -67,4 +71,69 @@ impl TransportErrorKind {
     pub const fn pubsub_unavailable() -> TransportError {
         RpcError::Transport(Self::PubsubUnavailable)
     }
+
+    /// Instantiate a new `TrasnportError::HttpError`.
+    pub const fn http_error(status: i64, body: String) -> TransportError {
+        RpcError::Transport(Self::HttpError(HTTPError { status, body }))
+    }
+}
+
+/// Type for holding HTTP errors such as 429 or -32005 by the RPC provider.
+#[derive(Debug, thiserror::Error)]
+#[error("HTTP error {status} with body: {body}")]
+pub struct HTTPError {
+    pub status: i64,
+    pub body: String,
+}
+
+impl HTTPError {
+    /// Analyzes the `status` and `body` to determine whether the request should be retried.
+    pub fn is_retry_err(&self) -> bool {
+        // alchemy throws it this way
+        if self.status == 429 {
+            return true;
+        }
+
+        // This is an infura error code for `exceeded project rate limit`
+        if self.status == -32005 {
+            return true;
+        }
+
+        // alternative alchemy error for specific IPs
+        if self.status == -32016 && self.body.contains("rate limit") {
+            return true;
+        }
+
+        // quick node error `"credits limited to 6000/sec"`
+        // <https://github.com/foundry-rs/foundry/pull/6712#issuecomment-1951441240>
+        if self.status == -32012 && self.body.contains("credits") {
+            return true;
+        }
+
+        // quick node rate limit error: `100/second request limit reached - reduce calls per second
+        // or upgrade your account at quicknode.com` <https://github.com/foundry-rs/foundry/issues/4894>
+        if self.status == -32007 && self.body.contains("request limit reached") {
+            return true;
+        }
+
+        match self.body.as_str() {
+            // this is commonly thrown by infura and is apparently a load balancer issue, see also <https://github.com/MetaMask/metamask-extension/issues/7234>
+            "header not found" => true,
+            // also thrown by infura if out of budget for the day and ratelimited
+            "daily request count exceeded, request rate limited" => true,
+            msg => {
+                msg.contains("rate limit")
+                    || msg.contains("rate exceeded")
+                    || msg.contains("too many requests")
+                    || msg.contains("credits limited")
+                    || msg.contains("request limit")
+            }
+        }
+    }
+}
+
+impl From<&ErrorPayload> for HTTPError {
```

Comment:
this seems like an inappropiate conversion, as it will produce invalid status codes on the HttpError